streaming/base/converters/dataframe_to_mds.py

```diff
@@ -0,0 +1,283 @@
+# Copyright 2023 MosaicML Streaming authors
+# SPDX-License-Identifier: Apache-2.0
+
+"""A utility to convert spark dataframe to MDS."""
+
+import json
+import logging
+import os
+import shutil
+from argparse import ArgumentParser, Namespace
+from collections.abc import Iterable
+from typing import Any, Callable, Dict, Iterable, Optional, Tuple, Union
+
+import pandas as pd
+from pyspark import TaskContext
+from pyspark.sql import SparkSession
+from pyspark.sql.dataframe import DataFrame
+from pyspark.sql.types import (ArrayType, BinaryType, BooleanType, ByteType, DateType,
+                               DayTimeIntervalType, DecimalType, DoubleType, FloatType,
+                               IntegerType, LongType, MapType, ShortType, StringType, StructField,
+                               StructType, TimestampNTZType, TimestampType)
+
+from streaming import MDSWriter
+from streaming.base.format.index import get_index_basename
+from streaming.base.format.mds.encodings import _encodings
+from streaming.base.storage.upload import CloudUploader
+
+logger = logging.getLogger(__name__)
+
+
+def is_iterable(obj: Any) -> bool:
+    """
+    Check if obj is iterable.
+    Args:
+        obj: python object
+    Return:
+        bool: true if obj is iterable false otherwise
+    """
+    return issubclass(type(obj), Iterable)
+
+
+def infer_dataframe_schema(dataframe: DataFrame,
+                           user_defined_cols: Optional[Dict[str, Any]] = None) -> Optional[Dict]:
+    """Retrieve schema to construct a dictionary or do sanity check for MDSWriter.
+
+    Args:
+        dataframe (spark dataframe): dataframe to inspect schema
+        user_defined_cols (Optional[Dict[str, Any]]): user specified schema for MDSWriter
+    Return:
+        If user_defined_cols is None, return schema_dict (dict): column name and dtypes that supported by MDSWriter
+        Else, return None
+    Exceptions:
+        Any of the datatype found to be unsupported by MDSWriter, then raise ValueError
+    """
+    mapping_spark_to_mds = {
+        ByteType: 'bytes',
+        ShortType: 'uint64',
+        IntegerType: 'int',
+        LongType: 'int64',
+        FloatType: 'float32',
+        DoubleType: 'float64',
+        DecimalType: 'str_decimal',
+        StringType: 'str',
+        BinaryType: 'bytes',
+        BooleanType: None,
+        TimestampType: None,
+        TimestampNTZType: None,
+        DateType: None,
+        DayTimeIntervalType: None,
+        ArrayType: None,
+        MapType: None,
+        StructType: None,
+        StructField: None
+    }
+
+    def map_spark_dtype(spark_data_type: Any):
+        for spark_dtype, mds_dtype in mapping_spark_to_mds.items():
+            if isinstance(spark_data_type, spark_dtype) and mds_dtype is not None:
+                return mds_dtype
+        raise ValueError(f'{spark_data_type} is not supported by MDSwriter')
```

Comment:
```suggestion
        mds_type = mapping_spark_to_mds.get(spark_dtype, None)
        if mds_type is None:
            raise ValueError(f'{spark_data_type} is not supported by MDSwriter')
        return mds_type
```
This should be way easier to read / more efficient.